sdk/schemaregistry/azure-data-schemaregistry-avro/src/main/java/com/azure/data/schemaregistry/avro/AvroSerializer.java

```diff
@@ -195,15 +216,66 @@ String getSchemaName(Object object) {
      * </ul>
      *
      * @param schemaBytes Bytes to read from.
+     *
      * @return true if the object has the single object payload header; false otherwise.
      *
      * @see <a href="https://avro.apache.org/docs/current/spec.html#single_object_encoding">Single Object Encoding</a>
      */
-    private static boolean isSingleObjectEncoded(byte[] schemaBytes) {
+    static boolean isSingleObjectEncoded(byte[] schemaBytes) {
         if (schemaBytes.length < V1_HEADER_LENGTH) {
             return false;
         }
 
         return V1_HEADER[0] == schemaBytes[0] && V1_HEADER[1] == schemaBytes[1];
     }
+
+    /**
+     * Gets a schema for the given class if it is an Avro primitive type.
+     *
+     * @param clazz Object class
+     *
+     * @return Matching primitive schema, otherwise {@code null} if it is not.
+     */
+    private static Schema getPrimitiveSchema(Class<?> clazz) {
+        final Schema schema = PRIMITIVE_SCHEMAS.get(clazz);
+        if (schema != null) {
+            return schema;
+        } else if (CharSequence.class.isAssignableFrom(clazz)) {
+            return PRIMITIVE_SCHEMAS.get(String.class);
+        } else if (ByteBuffer.class.isAssignableFrom(clazz)) {
+            return PRIMITIVE_SCHEMAS.get(Byte[].class);
+        } else {
+            return null;
+        }
+    }
+
+    /**
+     * Returns correct reader for decoding payload.
+     *
+     * @param writerSchema Avro schema fetched from schema registry store
+     *
+     * @return correct Avro DatumReader object given encoder configuration
+     */
+    @SuppressWarnings("unchecked")
+    private <T> DatumReader<T> getDatumReader(Schema writerSchema, TypeReference<T> typeReference) {
+        // Suppressing this warning because we know that the Type is a representation of the Class<T>
+        final Class<T> clazz = (Class<T>) typeReference.getJavaType();
+        final Schema primitiveSchema = getPrimitiveSchema(clazz);
+
+        if (primitiveSchema != null) {
+            assert primitiveSchema.getType().equals(writerSchema.getType());
```

Comment:
We shouldn't be using `assert` here as they are not enabled at runtime by default.